File: 03a/main.py
with open('input.txt', 'rt') as f:
    lines = (line.strip() for line in f.readlines())

cols = zip(*lines)
def most_common(it):
    it = list(it)
    ones = it.count('1')
    zeros = it.count('0')
    if ones > zeros:
        return '1'
    else:
        return '0'

gamma = ''.join(most_common(c) for c in cols)
gamma_int = int(gamma, base=2)

bin_len = len(gamma)
inverter = int('1' * bin_len, base=2)
epsilon_int = gamma_int ^ inverter

result = gamma_int * epsilon_int
print(result)

File: 03a/ruby/main.rb
FILENAME = "input.txt"
file = File.open(FILENAME)
lines = file.readlines.map(&:chomp).map { |line| line.split("") }
len = lines[0].length

gamma = []

for i in 0..(len - 1) do
  bits = lines.map { |line| line[i] }.map(&:to_i)
  if bits.count(1) > bits.count(0)
    gamma.append(1)
  else
    gamma.append(0)
  end
end
gamma = gamma.join("").to_i(base=2)
epsilon = gamma ^ (2 ** len - 1)

puts "gamma #{gamma}"
puts "epsilon #{epsilon}"
puts "result #{epsilon * gamma}"
